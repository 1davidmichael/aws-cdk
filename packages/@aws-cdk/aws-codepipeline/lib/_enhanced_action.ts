import { Aws, ResourceEnvironment } from '@aws-cdk/core';
import { IAction } from './action';

/**
 * Helper routines to work with Actions
 *
 * Can't put these on Action themselves since we only have an interface
 * and every library would need to reimplement everything (there is no
 * `ActionBase`).
 *
 * So here go the members that should have gone onto the Action class
 * but can't.
 *
 * It was probably my own idea but I don't want it anymore:
 * https://github.com/aws/aws-cdk/issues/10393
 */
export class EnhancedAction {
  constructor(private readonly action: IAction) {
  }

  /**
   * The environment this action runs in
   */
  public get env(): ResourceEnvironment {
    return {
      account: (this.action.actionProperties.role?.env.account
        ?? this.action.actionProperties?.resource?.env.account
        ?? this.action.actionProperties.account
        ?? Aws.ACCOUNT_ID),
      region: (this.action.actionProperties.resource?.env.region
        ?? this.action.actionProperties.region
        ?? Aws.REGION),
    };
  }

};