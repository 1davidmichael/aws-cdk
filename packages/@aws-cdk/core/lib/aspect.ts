import { IConstruct } from 'constructs';

/**
 * Represents an Aspect
 */
export interface IAspect {
  /**
   * All aspects can visit an IConstruct
   */
  visit(node: IConstruct): void;
}

const ASPECTS_SYMBOL = Symbol('cdk-aspects');

/**
 * Aspects can be applied to CDK tree scopes and can operate on the tree before
 * synthesis.
 */
export class Aspects {

  /**
   * Returns the `Aspects` object associated with a construct scope.
   * @param scope The scope for which these aspects will apply.
   */
  public static of(scope: IConstruct): Aspects {
    const aspects = (scope as any)[ASPECTS_SYMBOL];
    if (aspects) {
      return aspects;
    }

    Object.defineProperty(scope, ASPECTS_SYMBOL, {
      value: new Aspects(scope),
      configurable: false,
      enumerable: false,
    });

    return this.of(scope);
  }

  private readonly _aspects = new Array<IAspect>();

  private constructor(_construct: IConstruct) { }

  /**
   * Adds an aspect to be applied on this scope before synthesis.
   * @param aspect The aspect to apply.
   */
  public add(aspect: IAspect) {
    this._aspects.push(aspect);
  }

  /**
   * The list of aspects which were directly applied on this scope.
   */
  public get aspects(): IAspect[] {
    return  [ ...this._aspects ];
  }
}