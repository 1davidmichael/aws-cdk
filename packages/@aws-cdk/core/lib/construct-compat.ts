/**
 * Constructs compatibility layer.
 *
 * This file contains typescript aliases for types in "constructs" as part of
 * our migration to 2.x and in order to reduce the chance for merge conflicts as
 * we transition from the code base of 1.x to 2.x
 */

// /**
//  * Represents the building block of the construct graph.
//  *
//  * All constructs besides the root construct must be created within the scope of
//  * another construct.
//  */
// export class Construct extends constructs.Construct implements IConstruct {
//   /**
//    * Return whether the given object is a Construct
//    */
//   public static isConstruct(x: any): x is Construct {
//     return typeof x === 'object' && x !== null && CONSTRUCT_SYMBOL in x;
//   }

import * as constructs from 'constructs';

export type Construct = constructs.Construct;
export type IConstruct = constructs.IConstruct;
export type ConstructOrder = constructs.ConstructOrder;

export type DependableTrait = constructs.Dependable;
export type IDependable = constructs.IDependable;
export type ConstructNode = constructs.Node;
export type ConcreteDependable = constructs.DependencyGroup;
