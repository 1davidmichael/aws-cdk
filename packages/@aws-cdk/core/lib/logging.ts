import * as cxschema from '@aws-cdk/cloud-assembly-schema';
import * as cxapi from '@aws-cdk/cx-api';
import { IConstruct } from 'constructs';

/**
 * Logging methods for AWS CDK constructs.
 */
export class Logging {
  /**
   * Returns logging facilities associated with a construct.
   * @param construct The construct
   *
   * @example
   *
   * Logging.of(c).addWarning('this is a warning');
   *
   */
  public static of(construct: IConstruct): Logging {
    return new Logging(construct);
  }

  private constructor(private readonly construct: IConstruct) { }

  /**
   * Adds a { "aws:cdk:warning": <message> } metadata entry to this construct.
   * The toolkit will display the warning when an app is synthesized, or fail if
   * run in --strict mode. Stack trace will be included unless stack traces are
   * disabled for this scope.
   * @param message The warning message.
   */
  public addWarning(message: string) {
    this.addMessage(cxschema.ArtifactMetadataEntryType.WARN, message);
  }

  /**
   * Adds a { "aws:cdk:info": <message> } metadata entry to this construct.
   * The toolkit will display the info message when apps are synthesized.
   * Stack trace will be included unless stack traces are disabled for this scope.
   * @param message The info message.
   */
  public addInfo(message: string) {
    this.addMessage(cxschema.ArtifactMetadataEntryType.INFO, message);
  }

  /**
   * Adds an { "aws:cdk:error": <message> } metadata entry to this construct.
   * The toolkit will fail synthesis when errors are reported.
   * Stack trace will be included unless stack traces are disabled for this scope.
   * @param message The error message.
   */
  public addError(message: string) {
    this.addMessage(cxschema.ArtifactMetadataEntryType.ERROR, message);
  }

  private addMessage(key: string, message: string) {
    const stackTrace = this.construct.node.tryGetContext(cxapi.DISABLE_METADATA_STACK_TRACE) ? false : true;
    this.construct.node.addMetadata(key, message, { stackTrace });
  }
}