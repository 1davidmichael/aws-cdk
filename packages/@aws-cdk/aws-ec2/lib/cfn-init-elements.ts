import { Construct } from '@aws-cdk/core';
import * as s3 from '@aws-cdk/aws-s3';
import * as s3_assets from '@aws-cdk/aws-s3-assets';
import { isTagPropertyJson } from '@aws-cdk/cfnspec/lib/schema';

export enum InitRenderPlatform { WINDOWS, LINUX };
export enum InitElementType { PACKAGE, GROUP, USER, SOURCE, FILE, COMMAND, SERVICE };

export interface InitRenderOptions {
  /**
   * Which OS platform (Linux, Windows) are we rendering for. Impacts
   * which config types are available and how they are rendered.
   */
  platform: InitRenderPlatform;
  /**
   * Ordered index of current element type. Primarily used to auto-generate
   * command keys and retain ordering.
   */
  index: number;
}

/**
 * Base class for all CloudFormation Init elements
 */
export abstract class InitElement {

  /**
   * Returns the init element type for this element.
   */
  public abstract getElementType(): InitElementType;

  /**
   * Render the CloudFormation representation of this init element.
   *
   * @param options
   */
  public abstract renderElement(options: InitRenderOptions): any;

  /**
   *
   * @param scope
   */
  public abstract bind(scope: Construct): void;
}

/**
 * Options for InitCommand
 */
export interface InitCommandOptions {
  /**
   * Identifier key for this command
   *
   * You can use this to order commands, or to reference this command in a service
   * definition to restart a service after this command runs.
   *
   * @default - Automatically generated
   */
  readonly key?: string;

  /**
   * Sets environment variables for the command.
   *
   * This property overwrites, rather than appends, the existing environment.
   *
   * @default - Use current environment
   */
  readonly env?: Record<string, string>;

  /**
   * The working directory
   *
   * @default - Use default working directory
   */
  readonly cwd?: string;

  /**
   * Command to determine whether this command should be run
   *
   * If the test passes (exits with error code or %ERRORLEVEL% of 0), the command is run.
   *
   * @default - Always run this command
   */
  readonly test?: string;

  /**
   * Continue running if this command fails
   *
   * @default false
   */
  readonly ignoreErrors?: boolean;

  /**
   * Specifies how long to wait (in seconds) after a command has finished in case the command causes a reboot.
   *
   * A value of "forever" directs cfn-init to exit and resume only after the
   * reboot is complete. Set this value to 0 if you do not want to wait for
   * every command.
   *
   * For Windows systems only.
   *
   * @default "60"
   */
  readonly waitAfterCompletion?: string;
}

/**
 * Command to execute on the instance
 */
export class InitCommand extends InitElement {
  /**
   * Run a shell command
   *
   * You must escape the string appropriately.
   */
  public static shellCommand(shellCommand: string, options: InitCommandOptions = {}): InitCommand {
    return new InitCommand(shellCommand, options);
  }

  /**
   * Run a command from an argv array
   *
   * You do not need to escape space characters or enclose command parameters in quotes.
   */
  public static argvCommand(argv: string[], options: InitCommandOptions = {}): InitCommand {
    return new InitCommand(JSON.stringify(argv), options);
  }

  protected constructor(private readonly command: string, private readonly options: InitCommandOptions) {
    super();
  }

  public getElementType(): InitElementType {
    return InitElementType.COMMAND;
  }

  public renderElement(renderOptions: InitRenderOptions): any {
    const commandKey = this.options.key || (renderOptions.index + '').padStart(3, '0'); // 001, 005, etc.

    if (renderOptions.platform === InitRenderPlatform.LINUX && this.options.waitAfterCompletion !== undefined) {
      throw new Error('waitAfterCompletion is only valid for Linux systems.');
    }

    // TODO: Is there a more idiomatic way to do this?
    const renderedCommand = {} as Record<string, any>;
    renderedCommand[commandKey] = {
      command: this.command,
      env: this.options.env,
      cwd: this.options.cwd,
      test: this.options.test,
      ignoreErrors: this.options.ignoreErrors,
      waitAfterCompletion: this.options.waitAfterCompletion,
    };

    return renderedCommand;
  }

  /**
   * No-op for Commands
   */
  public bind(_scope: Construct): void {}

}

/**
 *
 * The content can be either inline in the template or the content can be
 * pulled from a URL.
 */
export class InitFileOptions {
  /**
   * The name of the owning group for this file.
   *
   * Not supported for Windows systems.
   *
   * @default 'root'
   */
  readonly group?: string;

  /**
   * The name of the owning user for this file.
   *
   * Not supported for Windows systems.
   *
   * @default 'root'
   */
  readonly owner?: string;

  /**
   * A six-digit octal value representing the mode for this file.
   *
   * Use the first three digits for symlinks and the last three digits for
   * setting permissions. To create a symlink, specify 120xxx, where xxx
   * defines the permissions of the target file. To specify permissions for a
   * file, use the last three digits, such as 000644.
   *
   * Not supported for Windows systems.
   *
   * @default '000644'
   */
  readonly mode?: string;
}

/**
 * Create files on the EC2 instance.
 */
export class InitFile extends InitElement {
  /**
   * Use a literal string as the file content
   */
  public static fromString(fileName: string, s: string, options: InitFileOptions = {}): InitFile {
  }

  /**
   * Write a symlink with the given symlink target
   */
  public static symlink(fileName: string, target: string, options: InitFileOptions = {}): InitFile {
  }

  /**
   * Use a base64 string as the file content
   */
  public static fromBase64String(fileName: string, base64: string, options: InitFileOptions = {}): InitFile {
  }

  /**
   * Use a JSON object as the file content, write it to a JSON file.
   *
   * May contain tokens.
   */
  public static fromJsonObject(fileName: string, obj: Record<string, any>, options: InitFileOptions = {}): InitFile {
  }

  /**
   * Read a file from disk and use its contents
   *
   * The file will be embedded in the template, so care should be taken to not
   * exceed the template size.
   *
   * The file will be base64-encoded if it contains nonprintable characters.
   */
  public static fromFileInline(targetFileName: string, sourceFileName: string, options: InitFileOptions = {}): InitFile {
  }

  /**
   * Create an asset from the given file and use that
   *
   * This is appropriate for files that are too large to embed into the template.
   */
  public static fromFileAsset(targetFileName: string, sourceFileName: string, options: InitFileOptions = {}): InitFile {
  }

  /**
   * Download from a URL at instance startup time
   */
  public static fromUrl(targetFileName: string, url: string, options: InitFileOptions = {}): InitFile {
  }

  /**
   * Use a file from an asset at instance startup time
   */
  public static fromAsset(targetFileName: string, asset: s3_assets.Asset, options: InitFileOptions = {}): InitFile {
  }

  protected constructor() {
    super();
  }
}

/**
 * Create Linux/UNIX groups and to assign group IDs.
 *
 * Not supported for Windows systems.
 */
export class InitGroup extends InitElement {

  /**
   * Map a group name to a group id
   */
  public static fromName(groupName: string, groupId?: number): InitGroup {
    return new InitGroup(groupName, groupId);
  }

  protected constructor(private groupName: string, private groupId?: number) {
    super();
  }

  public getElementType(): InitElementType { return InitElementType.GROUP; }

  public renderElement(options: InitRenderOptions): any {
    if (options.platform === InitRenderPlatform.WINDOWS) {
      throw new Error('Init groups are not supported on Windows');
    }

    // TODO: Is there a more idiomatic way to do this?
    const element = {} as Record<string, any>;
    element[this.groupName] = this.groupId !== undefined ? { gid: this.groupId } : {};
    return element;
  }

  // No-op
  public bind(_scope: Construct): void {}
}

/**
 * Create Linux/UNIX users and to assign user IDs.
 *
 * Users are created as non-interactive system users with a shell of
 * /sbin/nologin. This is by design and cannot be modified.
 *
 * Not supported for Windows systems.
 */
export class InitUser extends InitElement {
  /**
   * Map a user name to a user id
   */
  public static fromName(userName: string, uid: number, groups?: string[], homeDir?: string): InitUser {
    return new InitUser(userName, uid, groups, homeDir);
  }

  private readonly userName: string;
  private readonly uid: number;
  private readonly groups?: string[];
  private readonly homeDir?: string;

  protected constructor(userName: string, uid: number, groups?: string[], homeDir?: string) {
    super();

    this.userName = userName;
    this.uid = uid;
    this.groups = groups;
    this.homeDir = homeDir;
  }

  public getElementType(): InitElementType { return InitElementType.USER; }

  public renderElement(options: InitRenderOptions): any {
    if (options.platform === InitRenderPlatform.WINDOWS) {
      throw new Error('Init users are not supported on Windows');
    }

    // TODO: Is there a more idiomatic way to do this?
    const element = {} as Record<string, any>;
    element[this.userName] = {
      uid: this.uid,
      groups: this.groups,
      homeDir: this.homeDir,
    };
    return element;
  }

  // No-op
  public bind(_scope: Construct): void {}
}

/**
 * A package to be installed during cfn-init time
 */
export class InitPackage extends InitElement {
  /**
   * Install an RPM from an HTTP URL or a location on disk
   */
  public static rpm(location: string, key?: string): InitPackage {
  }

  /**
   * Install a package using Yum
   */
  public static yum(packageName: string, ...versions: string[]): InitPackage {
  }

  /**
   * Install a package from RubyGems
   */
  public static rubyGem(gemName: string, ...versions: string[]): InitPackage {
  }

  /**
   * Install a package from PyPI
   */
  public static python(packageName: string, ...versions: string[]): InitPackage {
  }

  /**
   * Install a package using APT
   */
  public static apt(packageName: string, ...versions: string[]): InitPackage {
  }

  /**
   * Install an MSI package from an HTTP URL or a location on disk
   */
  public static msi(location: string, key?: string): InitPackage {
  }

  protected constructor() {
    super();
  }
}

/**
 * A services that be enabled, disabled or restarted when the instance is launched.
 */
export class InitService extends InitElement {
  /**
   * Enable and start the given service, optionally restarting it
   */
  public static enable(serviceName: string, options: InitServiceOptions = {}): InitService {
  }

  /**
   * Disable and stop the given service, optionally restarting it
   */
  public static disable(serviceName: string, options: InitServiceOptions = {}): InitService {
  }

  /**
   * Leave the running state of the service alone, optionally restarting it
   */
  public static restart(serviceName: string, options: InitServiceOptions = {}): InitService {
  }
}

export interface InitServiceOptions {
  /**
   * Enable or disable this service
   *
   * Set to true to ensure that the service will be started automatically upon boot.
   *
   * Set to false to ensure that the service will not be started automatically upon boot.
   *
   * @default - true if used in `InitService.enable()`, false if used in `InitService.disable()`,
   * no change to service state if used in `InitService.restart()`.
   */
  readonly enabled?: boolean;

  /**
   * Make sure this service is running or not running after cfn-init finishes.
   *
   * Set to true to ensure that the service is running after cfn-init finishes.
   *
   * Set to false to ensure that the service is not running after cfn-init finishes.
   *
   * @default - Same value as 'enabled'
   * no change to service state if used in `InitService.restart()`.
   */
  readonly ensureRunning?: boolean;

  /**
   * Restart service if cfn-init touches one of these files
   *
   * Only works for files listed in the `files` section.
   *
   * @default - No files trigger restart
   */
  readonly restartAfterFiles?: string[];

  /**
   * Restart service if cfn-init expands an archive into one of these directories.
   *
   * @default - No sources trigger restart
   */
  readonly restartAfterSources?: string[];

  /**
   * Restart service if cfn-init installs or updates one of these packages
   *
   * A map of package manager to list of package names.
   *
   * @default - No packages trigger restart
   */
  readonly restartAfterPackages?: Record<string, string[]>;

  /**
   * Restart service after cfn-init runs the given command
   *
   * Takes a list of command names.
   *
   * @default - No commands trigger restart
   */
  readonly restartAfterCommands?: string[];
}

/**
 * Extract an archive into a directory
 */
export class InitSource extends InitElement {
  /**
   * Retrieve a URL and extract it into the given directory
   */
  public static fromUrl(targetDirectory: string, url: string): InitSource {
    return new InitSource(targetDirectory, url);
  }

  /**
   * Extract a GitHub branch into a given directory
   */
  public static fromGitHub(targetDirectory: string, owner: string, repo: string, refSpec?: string): InitSource {
    return InitSource.fromUrl(targetDirectory, `https://github.com/${owner}:${repo}/tarball/${refSpec ?? 'master'}`);
  }

  /**
   * Extract an archive stored in an S3 bucket into the given directory
   */
  public static fromS3Object(targetDirectory: string, bucket: s3.IBucket, key: string) {
    return InitSource.fromUrl(targetDirectory, bucket.urlForObject(key));
  }

  /**
   * Create an asset from the given directory and use that
   */
  public static fromDirectoryAsset(targetDirectory: string, sourceDirectory: string): InitSource {
    throw new Error('Not implemented');
  }

  /**
   * Extract a diretory from an existing directory asset
   */
  public static fromAsset(targetDirectory: string, asset: s3_assets.Asset): InitSource {
    throw new Error('Not implemented');
  }

  private readonly targetDirectory: string;
  private readonly url: string;

  protected constructor(targetDirectory: string, url: string) {
    super();

    this.targetDirectory = targetDirectory;
    this.url = url;
  }

  public getElementType(): InitElementType { return InitElementType.SOURCE; }

  public renderElement(_options: InitRenderOptions): any {
    const element = {} as Record<string, any>;
    element[this.targetDirectory] = this.url;
    return element;
  }

  public bind(_scope: Construct): void {
    throw new Error('Not yet implemented');
  }
}