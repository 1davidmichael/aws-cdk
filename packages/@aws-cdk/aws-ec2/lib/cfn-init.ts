import * as iam from '@aws-cdk/aws-iam';
import { Aws, CfnResource, Construct } from '@aws-cdk/core';
import * as crypto from 'crypto';
import { InitBindOptions, InitElement, InitElementType, InitRenderOptions, InitRenderPlatform } from './cfn-init-elements';
import { UserData } from './user-data';

/**
 * A CloudFormation-init configuration
 */
export class CloudFormationInit {
  /**
   * Build a new config from a set of Init Elements
   */
  public static fromElements(...elements: InitElement[]): CloudFormationInit {
    return CloudFormationInit.fromConfig(new InitConfig(elements));
  }

  /**
   * Use an existing InitConfig object as the default and only config
   */
  public static fromConfig(config: InitConfig): CloudFormationInit {
    return CloudFormationInit.fromConfigSets({
      configSets: {
        default: ['config'],
      },
      configs: { config },
    });
  }

  /**
   * Build a CloudFormationInit from config sets
   */
  public static fromConfigSets(props: ConfigSetProps): CloudFormationInit {
    return new CloudFormationInit(props.configSets, props.configs);
  }

  private readonly _configSets: Record<string, string[]> = {};
  private readonly _configs: Record<string, InitConfig> = {};

  protected constructor(configSets: Record<string, string[]>, configs: Record<string, InitConfig>) {
    Object.assign(this._configSets, configSets);
    Object.assign(this._configs, configs);
  }

  /**
   * Add a config with the given name to this CloudFormationInit object
   */
  public addConfig(configName: string, config: InitConfig) {
    if (this._configs[configName]) {
      throw new Error(`CloudFormationInit already contains a config named '${configName}'`);
    }
    this._configs[configName] = config;
  }

  /**
   * Add a config set with the given name to this CloudFormationInit object
   *
   * The new configset will reference the given configs in the given order.
   */
  public addConfigSet(configSetName: string, configNames: string[] = []) {
    if (this._configSets[configSetName]) {
      throw new Error(`CloudFormationInit already contains a configSet named '${configSetName}'`);
    }

    const unk = configNames.filter(c => !this._configs[c]);
    if (unk.length > 0) {
      throw new Error(`Unknown configs referenced in definition of '${configSetName}': ${unk}`);
    }

    this._configSets[configSetName] = [...configNames];
  }

  /**
   * Attach the CloudFormation Init config to the given resource
   *
   * This returns an `AttachedCloudFormationInit` object, which can be used to apply
   * the init to one or more instances and/or autoscaling groups. Applying the
   * `AttachedCloudFormationInit` does NOT increase the ResourceSignalCount for
   * the attached resource, you have to do that separately.
   *
   * You only need to use this API if want to reuse the same init config for
   * multiple resources. If not, `instance.applyCloudFormationInit()` or
   * `autoScalingGroup.applyCloudFormationInit()` achieve the same result in a
   * simpler way.
   */
  public attach(definingResource: CfnResource, attachOptions: AttachInitOptions): AttachedCloudFormationInit {
    // FIXME: This this eagerly renders, it will not reflect mutations made after attaching.
    // Is that okay?
    const configData = this.renderInit(attachOptions.platform);
    definingResource.addMetadata('AWS::CloudFormation::Init', configData);
    const fingerprint = contentHash(JSON.stringify(configData)).substr(0, 16);

    const self = this;

    // Anonymous subclass of an abstract class because I don't want users to be able to
    // instantiate this. It is a type that represents an action you took.
    // tslint:disable-next-line:new-parens
    return new class extends AttachedCloudFormationInit {
      public apply(consumingResource: CfnResource, useOptions: ApplyInitOptions): void {
        self.bind(consumingResource, { instanceRole: useOptions.instanceRole });
        useOptions.instanceRole.addToPolicy(new iam.PolicyStatement({
          actions: ['cloudformation:DescribeStackResource', 'cloudformation:SignalResource'],
          resources: [Aws.STACK_ID],
        }));

        // To identify the resources that (a) have the metadata and (b) where the signal needs to be sent
        // (may be different), we need { region, stackName, logicalId }
        const initLocator = `--region ${Aws.REGION} --stack ${Aws.STACK_NAME} --resource ${definingResource.logicalId}`;
        const signalLocator = `--region ${Aws.REGION} --stack ${Aws.STACK_NAME} -resource ${consumingResource.logicalId}`;
        const configSets = (useOptions.configSets ?? ['default']).join(',');

        if (useOptions.embedFingerprint ?? true) {
          // It just so happens that the comment char is '#' for both bash and PowerShell
          useOptions.userData.addCommands(`# fingerprint: ${fingerprint}`);
        }

        if (attachOptions.platform === InitRenderPlatform.WINDOWS) {
          useOptions.userData.addCommands(
            `cfn-init.exe -v ${initLocator} -c ${configSets}`,
            `cfn-signal.exe -e %ERRORLEVEL% ${signalLocator}`,
          );
        } else {
          useOptions.userData.addCommands(
            // Run a subshell without 'errexit', so we can signal using the exit code of cfn-init
            '(',
            '  set +e',
            `  /opt/aws/bin/cfn-init -v ${initLocator} -c ${configSets}`,
            `  /opt/aws/bin/cfn-signal -e $? ${signalLocator}`,
            ')',
          );
        }
      }
    };
  }

  private bind(scope: Construct, options: InitBindOptions) {
    for (const config of Object.values(this._configs)) {
      config.bind(scope, options);
    }
  }

  private renderInit(platform: InitRenderPlatform): any {
    const nonEmptyConfigs = mapValues(this._configs, c => c.isEmpty ? undefined : c);

    return {
      configSets: mapValues(this._configSets, configNames => configNames.filter(name => nonEmptyConfigs[name] !== undefined)),
      ...mapValues(nonEmptyConfigs, c => c.renderConfig(platform)),
    };
  }
}

/**
 * A CloudFormationInit object that has been attached to a Resource
 *
 * Can be applied to an instance, role and UserData.
 */
export abstract class AttachedCloudFormationInit {
  /**
   * Apply the config to a resource
   */
  public abstract apply(resource: CfnResource, options: ApplyInitOptions): void;
}

/**
 * Options for attach a CloudFormationInit to a resource
 */
export interface AttachInitOptions {
  /**
   * Platform to render for
   */
  readonly platform: InitRenderPlatform;
}

/**
 * Options for applying a CloudFormationInit to an arbitrary resource
 */
export interface ApplyInitOptions {
  /**
   * Instance role of the consuming instance or fleet
   */
  readonly instanceRole: iam.IRole;

  /**
   * UserData to add commands to
   */
  readonly userData: UserData;

  /**
   * ConfigSet to activate
   *
   * @default ['default']
   */
  readonly configSets?: string[];

  /**
   * Whether to embed a hash into the userData
   *
   * If `true` (the default), a hash of the config will be embedded into the
   * UserData, so that if the config changes, the UserData changes and
   * the instance will be replaced.
   *
   * If `false`, no such hash will be embedded, and if the CloudFormation Init
   * config changes nothing will happen to the running instance.
   *
   * @default true
   */
  readonly embedFingerprint?: boolean;
}

/**
 * A collection of configuration elements
 */
export class InitConfig {
  private readonly elements = new Array<InitElement>();

  constructor(elements: InitElement[]) {
    this.add(...elements);
  }

  /**
   * Whether this configset has elements or not
   */
  public get isEmpty() {
    return this.elements.length === 0;
  }

  /**
   * Add one or more elements to the config
   */
  public add(...elements: InitElement[]) {
    this.elements.push(...elements);
  }

  /**
   * Called when the config is applied to an instance
   */
  public bind(scope: Construct, options: InitBindOptions) {
    for (const element of this.elements) {
      element.bind(scope, options);
    }
  }

  /**
   * Render the config
   */
  public renderConfig(platform: InitRenderPlatform): any {
    const renderOptions = { platform };

    return {
      packages: this.renderConfigForType(InitElementType.PACKAGE, renderOptions),
      groups: this.renderConfigForType(InitElementType.GROUP, renderOptions),
      users: this.renderConfigForType(InitElementType.USER, renderOptions),
      sources: this.renderConfigForType(InitElementType.SOURCE, renderOptions),
      files: this.renderConfigForType(InitElementType.FILE, renderOptions),
      commands: this.renderConfigForType(InitElementType.COMMAND, renderOptions),
      services: this.renderConfigForType(InitElementType.SERVICE, renderOptions),
    };
  }

  private renderConfigForType(elementType: InitElementType, renderOptions: Omit<InitRenderOptions, 'index'>): Record<string, any> | undefined {
    const elements = this.elements.filter(elem => elem.elementType === elementType);
    if (elements.length === 0) { return undefined; }

    const ret: Record<string, any> = {};
    elements.forEach((elem, index) => {
      deepMerge(ret, elem.renderElement({ index, ...renderOptions }));
    });
    return ret;
  }
}

/**
 * Options for CloudFormationInit.withConfigSets
 */
export interface ConfigSetProps {
  /**
   * The definitions of each config set
   */
  readonly configSets: Record<string, string[]>;

  /**
   * The sets of configs to pick from
   */
  readonly configs: Record<string, InitConfig>;
}

function deepMerge(target: Record<string, any>, src: Record<string, any>) {
  for (const [key, value] of Object.entries(src)) {
    if (typeof value === 'object' && value) {
      target[key] = {};
      deepMerge(target[key], value);
    } else {
      target[key] = value;
    }
  }
}

/**
 * Map a function over values of an object
 *
 * If the mapping function returns undefined, remove the key
 */
function mapValues<A, B>(xs: Record<string, A>, fn: (x: A) => B | undefined): Record<string, B> {
  const ret: Record<string, B> = {};
  for (const [k, v] of Object.entries(xs)) {
    const mapped = fn(v);
    if (mapped !== undefined) {
      ret[k] = mapped;
    }
  }
  return ret;
}

function contentHash(content: string) {
  return crypto.createHash('sha256').update(content).digest('hex');
}