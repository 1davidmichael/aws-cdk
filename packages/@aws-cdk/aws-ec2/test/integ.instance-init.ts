#!/usr/bin/env node
import * as ec2 from '../lib';
import * as cdk from '@aws-cdk/core';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'integ-init', {
  env: { account: process.env.CDK_DEFAULT_ACCOUNT, region: process.env.CDK_DEFAULT_REGION },
});

const vpc = ec2.Vpc.fromLookup(stack, 'VPC', { isDefault: true });

new ec2.Instance(stack, 'Instance2', {
  vpc,
  vpcSubnets: { subnetType: ec2.SubnetType.PUBLIC },
  instanceType: ec2.InstanceType.of(ec2.InstanceClass.BURSTABLE2, ec2.InstanceSize.MICRO),
  machineImage: new ec2.AmazonLinuxImage(),
  userDataCausesReplacement: true,
  init: ec2.CloudFormationInit.fromElements(
    ec2.InitCommand.shellCommand('echo hello2'),
  ),
});

app.synth();