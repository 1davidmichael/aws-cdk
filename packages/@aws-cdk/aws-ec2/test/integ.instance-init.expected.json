{
  "Resources": {
    "InstanceInstanceSecurityGroupF0E2D5BE": {
      "Type": "AWS::EC2::SecurityGroup",
      "Properties": {
        "GroupDescription": "integ-init/Instance/InstanceSecurityGroup",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1"
          }
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "integ-init/Instance"
          }
        ],
        "VpcId": "vpc-60900905"
      }
    },
    "InstanceInstanceRoleE9785DE5": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::Join": [
                    "",
                    [
                      "ec2.",
                      {
                        "Ref": "AWS::URLSuffix"
                      }
                    ]
                  ]
                }
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "integ-init/Instance"
          }
        ]
      }
    },
    "InstanceInstanceRoleDefaultPolicy4ACE9290": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "cloudformation:DescribeStackResource",
                "cloudformation:SignalResource"
              ],
              "Effect": "Allow",
              "Resource": {
                "Ref": "AWS::StackId"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "InstanceInstanceRoleDefaultPolicy4ACE9290",
        "Roles": [
          {
            "Ref": "InstanceInstanceRoleE9785DE5"
          }
        ]
      }
    },
    "InstanceInstanceProfileAB5AEF02": {
      "Type": "AWS::IAM::InstanceProfile",
      "Properties": {
        "Roles": [
          {
            "Ref": "InstanceInstanceRoleE9785DE5"
          }
        ]
      }
    },
    "InstanceC1063A87": {
      "Type": "AWS::EC2::Instance",
      "Properties": {
        "AvailabilityZone": "us-east-1a",
        "IamInstanceProfile": {
          "Ref": "InstanceInstanceProfileAB5AEF02"
        },
        "ImageId": {
          "Ref": "SsmParameterValueawsserviceamiamazonlinuxlatestamznamihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter"
        },
        "InstanceType": "t2.micro",
        "SecurityGroupIds": [
          {
            "Fn::GetAtt": [
              "InstanceInstanceSecurityGroupF0E2D5BE",
              "GroupId"
            ]
          }
        ],
        "SubnetId": "subnet-e19455ca",
        "Tags": [
          {
            "Key": "Name",
            "Value": "integ-init/Instance"
          }
        ],
        "UserData": {
          "Fn::Base64": {
            "Fn::Join": [
              "",
              [
                "#!/bin/bash\n# fingerprint: 7c06760dd0086233\n(\n  set +e\n  /opt/aws/bin/cfn-init -v --region ",
                {
                  "Ref": "AWS::Region"
                },
                " --stack ",
                {
                  "Ref": "AWS::StackName"
                },
                " --resource InstanceC1063A87 -c default\n  /opt/aws/bin/cfn-signal -e $? --region ",
                {
                  "Ref": "AWS::Region"
                },
                " --stack ",
                {
                  "Ref": "AWS::StackName"
                },
                " --resource InstanceC1063A87\n)"
              ]
            ]
          }
        }
      },
      "DependsOn": [
        "InstanceInstanceRoleDefaultPolicy4ACE9290",
        "InstanceInstanceRoleE9785DE5"
      ],
      "CreationPolicy": {
        "ResourceSignal": {
          "Count": 1,
          "Timeout": "PT5M"
        }
      },
      "Metadata": {
        "AWS::CloudFormation::Init": {
          "configSets": {
            "default": [
              "config"
            ]
          },
          "config": {
            "commands": {
              "000": {
                "command": "echo hello"
              }
            }
          }
        }
      }
    }
  },
  "Parameters": {
    "SsmParameterValueawsserviceamiamazonlinuxlatestamznamihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>",
      "Default": "/aws/service/ami-amazon-linux-latest/amzn-ami-hvm-x86_64-gp2"
    }
  }
}